### ISSUE TITLE
IOError on synced state module causes memory leak during state run

### ISSUE THREAD

**Issue by jodv** [saltstack/salt#32003]
### Description of Issue/Question

When executing `state.sls` on a host, we saw the job process bloat to 60GB of VMS. Upon investigation we found that one of the custom states on the host was causing a Python IOError when being read from.  This caused `saltutil.sync_states` to error out, but caused a `state.sls` execution to hang indefinitely and consume all available memory.

Obviously the file/storage being in a bad state is the real problem, but the Lazy Loader shouldn't endlessly spin and bloat when faced with that condition.
### Steps to Reproduce Issue

Here is the traceback returned from running `saltutil.sync_states` with the problematic file in `extmods/states`

```
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.6/site-packages/salt/minion.py", line 1146, in _thread_return
        return_data = func(*args, **kwargs)
      File "/usr/lib/python2.6/site-packages/salt/modules/saltutil.py", line 303, in sync_states
        ret = _sync('states', saltenv)
      File "/usr/lib/python2.6/site-packages/salt/modules/saltutil.py", line 129, in _sync
        dst_digest = salt.utils.get_hash(dest, hash_type)
      File "/usr/lib/python2.6/site-packages/salt/utils/__init__.py", line 1727, in get_hash
        for chunk in iter(lambda: ifile.read(chunk_size), b''):
      File "/usr/lib/python2.6/site-packages/salt/utils/__init__.py", line 1727, in <lambda>
        for chunk in iter(lambda: ifile.read(chunk_size), b''):
    IOError: [Errno 5] Input/output error
```

It's not just Python that was having issues with the file

```
# md5sum extmods/states/git.py
md5sum: extmods/states/git.py: Input/output error
```

I don't know how to tell you to replicate an IOError unfortunately.  Usually, it's indicative of a hardware issue AFAIK.  But once the file was in that state, any state.sls run that involved this state module being lazyloaded resulted in the bad behavior.
### Versions Report

Yes, I know most of these versions are outdated, to put it nicely. :turtle: 

```
$ salt --versions-report
           Salt: 2015.5.2-3263-g87313fc
         Python: 2.6.6 (r266:84292, Nov 21 2013, 10:50:32)
         Jinja2: 2.7-dev
       M2Crypto: 0.20.2
 msgpack-python: 0.4.6
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 14.3.0
           RAET: Not Installed
            ZMQ: 3.2.4
           Mako: 0.3.4
```

**Comment by jfindlay** [comment]
@jodv, thanks for reporting.  Is it possible to reproduce this on the latest version of salt?  I'm not sure how to simulate an input/output error on a file from storage, but if you're not able to upgrade, I'll see what I can do.

**Comment by cachedout** [comment]
I'm not sure at all why upgrading would solve this issue.

This is sort of a tough one because I tend to wonder if the IOError is something that got produced _after_ the storage layer delivered some type of data to the OS. Maybe garbage data? We do read that file in chunks. My shot-in-the-dark-guess here is that we got a stream of data from a failing storage layer that we dutifully read. Though, we should be reading in chunks anyway, so that's odd..

This will be difficult to debug. I totally believe that it happened but without a way to inspect the stack at the time that it's happening, I'm at a bit of a loss...

**Comment by jodv** [comment]
I fully understand that reproducing/debugging this will be difficult, if not impossible.  I haven't gotten time to try yet, but I intend to when I can.  Unfortunately, the issue was blocking something such that I had to remediate quickly and did not get a core file.  This may not even be Salt, it may be something in the CPython interpreter.

In any case, I filed the bug for due diligence purposes and if I encounter any more instances of this, I'll do my best to get more data. :smile:
